lib/phoenix/endpoint/cowboy_handler.ex

```diff
@@ -86,17 +86,35 @@ defmodule Phoenix.Endpoint.CowboyHandler do
   defp default_for(Phoenix.Transports.WebSocket), do: Phoenix.Endpoint.CowboyWebSocket
   defp default_for(_), do: nil
 
+  # Bake in retrieving the actual port from the adapter if possible, otherwise
+  # fallback to lower-level/less portable behavior.
+  defmacrop get_listener_port(ref) do
+    case :info in Plug.Adapters.Cowboy.__info__(:functions) do
+      true  -> quote do: unquote(ref) |> Plug.Adapters.Cowboy.info()[:port]
+      false -> quote do: unquote(ref) |> :ranch.get_port
+    end
+  end
+
   @doc """
   Callback to start the Cowboy endpoint.
   """
-  def start_link(scheme, endpoint, config, {m, f, a}) do
+  def start_link(_scheme, endpoint, config, {m, f, a}) do
     case apply(m, f, a) do
       {:ok, pid} ->
-        Logger.info info(scheme, endpoint, config)
+        listen_port = a |> hd |> get_listener_port()
```

Comment:
endpoint.url != scheme/endpoint being started. The same endpoint can be started under 5 different host + scheme combination but the canonical url is just one.